### ISSUE TITLE
PDM Score is Zero

### ISSUE THREAD

**Issue by hongxiaoy** [autonomousvision/navsim#28]
When I run the evaluation scripts of human agent on navtest splits, though I didn't download all the logs, the score of each scenario is 0, except the NC and TTC, which cause the final PDMS is 0. And I can't where is wrong. Could you help me with it? Thanks.

**Comment by hongxiaoy** [comment]
![image](https://github.com/user-attachments/assets/da30170a-0a06-4197-b350-836ca30af307)

**Comment by DanielDauner** [comment]
Hi @hongxiaoy 

That is a strange bug and hasn’t occurred previously. 

The evaluation only considers 3656 out of 12146 scenarios in `navtest`. You also mentioned that you didn’t download all the logs from the test split. Could you try:

1. Downloading all the available logs in the [test split](https://github.com/autonomousvision/navsim/blob/main/download/download_test.sh).
2. Creating a [new metric cache](https://github.com/autonomousvision/navsim/blob/main/docs/cache.md).
3. Re-running the [evaluation](https://github.com/autonomousvision/navsim/blob/main/scripts/evaluation/run_human_agent_pdm_score_evaluation.sh)?

Please note that for (2+3) you need to set the argument `train_test_split=navtest`. You also need to re-run (2) if you recently updated from `v1.0` to `v1.1`.

I hope that helps!
Daniel

**Comment by hongxiaoy** [comment]
Thanks for your quick response, I'll follow your instructions and try again.

**Comment by hongxiaoy** [comment]
> Hi @hongxiaoy
> 
> That is a strange bug and hasn’t occurred previously. 
> 
> The evaluation only considers 3656 out of 12146 scenarios in `navtest`. You also mentioned that you didn’t download all the logs from the test split. Could you try:
> 
> 1. Downloading all the available logs in the [test split](https://github.com/autonomousvision/navsim/blob/main/download/download_test.sh).
> 2. Creating a [new metric cache](https://github.com/autonomousvision/navsim/blob/main/docs/cache.md).
> 3. Re-running the [evaluation](https://github.com/autonomousvision/navsim/blob/main/scripts/evaluation/run_human_agent_pdm_score_evaluation.sh)?
> 
> Please note that for (2+3) you need to set the argument `train_test_split=navtest`. You also need to re-run (2) if you recently updated from `v1.0` to `v1.1`.
> 
> I hope that helps! Daniel

I have tried, but still not work.
![image](https://github.com/user-attachments/assets/11187fc6-d11e-49a3-9321-f89c82c80dd1)

**Comment by hongxiaoy** [comment]
And this is the data frame for PDMS:
![image](https://github.com/user-attachments/assets/75745d65-227b-40ee-8b1c-b148c771acdc)
It seems that NC and TTC are normal, others are weird.

**Comment by DanielDauner** [comment]
Strange... At least the number of scenarios is correct now. 
Have you installed the maps correctly? Are there any changes you made to the code?

**Comment by hongxiaoy** [comment]
I just download this repo and follow the install.md to prepare dataset, then run the evaluation script for human agent.

**Comment by hongxiaoy** [comment]
The env variables are:
![image](https://github.com/user-attachments/assets/e54a2f76-7bcb-4946-bf88-a670ecfa0a47)
and maps dir are like this:
![image](https://github.com/user-attachments/assets/dad844ae-97fd-4eb3-9650-f4099da2bbd3)

**Comment by DanielDauner** [comment]
Okay strange. Could you maybe check if the environment variables are set correctly. In the above example, it looks like 
```
export NUPLAN_MAPS_ROOT="/mnt/vdb1/hongxiao.yu/data/navsim/maps"
```
is the correct variable. Please also check if the remaining variables are set correctly. Given the error you're having, I assume there is a problem when the metric cache is created. You might need to re-run the metric caching during debugging.

**Comment by hongxiaoy** [comment]
Not work. I also found that the ego_mlp agent can generate a normal PDMS.

**Comment by hongxiaoy** [comment]
> Okay strange. Could you maybe check if the environment variables are set correctly. In the above example, it looks like
> 
> ```
> export NUPLAN_MAPS_ROOT="/mnt/vdb1/hongxiao.yu/data/navsim/maps"
> ```
> 
> is the correct variable. Please also check if the remaining variables are set correctly. Given the error you're having, I assume there is a problem when the metric cache is created. You might need to re-run the metric caching during debugging.
> 
> Edit: Maybe also re-install the maps folder.

I changed a machine to run the evaluation, then the PDMS is 0.96, and every metric item is normal. Anyway, thank you for your suggestion again. 👍

**Comment by hongxiaoy** [comment]
After the comparison, I found that the bug comes from `np.linalg.pinv()`, which generate different output given the same input. If I use the `torch.linalg.inv()`, it's normal.

**Comment by DanielDauner** [comment]
That's interesting and hasn't previously occurred. Thank you very much for the follow-up. If this error occurs more frequently, we will consider adapting the implementation!

**Comment by ztony0712** [comment]
> That's interesting and hasn't previously occurred. Thank you very much for the follow-up. If this error occurs more frequently, we will consider adapting the implementation!

We also encounter this problem. It was solved by changing to another computer, as well.

**Comment by DanielDauner** [comment]
@ztony0712 @hongxiaoy Any chance you could forward me the `numpy` version where the code works or doesn't run? Thanks!

**Comment by hongxiaoy** [comment]
> @ztony0712 @hongxiaoy Any chance you could forward me the `numpy` version where the code works or doesn't run? Thanks!

Of course.

NumPy version: `1.23.4`

Bug Appears at:
- https://github.com/autonomousvision/navsim/blob/2623b40b587d93ac4d66b54594cc0f33a9e0051a/navsim/planning/simulation/planner/pdm_planner/simulation/batch_lqr_utils.py#L124
- https://github.com/autonomousvision/navsim/blob/2623b40b587d93ac4d66b54594cc0f33a9e0051a/navsim/planning/simulation/planner/pdm_planner/simulation/batch_lqr_utils.py#L179

Solution:
- change to `torch.linalg.pinv(torch.tensor(batch_matmul(A_T, A) + jerk_penalty * batch_matmul(R_T, R))).numpy(), A_T`
- change to `intermediate = batch_matmul(torch.linalg.pinv(torch.tensor(batch_matmul(A_T, A) + Q)).numpy(), A_T)`

**Comment by ztony0712** [comment]
> @ztony0712 @hongxiaoy Any chance you could forward me the `numpy` version where the code works or doesn't run? Thanks!

numpy                     1.23.4
The numpy version satisfies your requirement, but it doesn't work.
I change np.linalg.pinv() to torch.linalg.inv() and it works. Thanks to @hongxiaoy !

**Comment by RogerChern** [comment]
I guess this might be some micro kernel selection issues for openblas which happens from time to time(e.g. https://github.com/numpy/numpy/issues/22502) when running codes on CPUs with newer arch like intel 8457C.

Just bump my numpy to 1.26.4 on the linux machine solve everything.

Besides, purge all metric_caches and rebuild them after upgrade your numpy.
